### ISSUE TITLE
Memory Leak when downloading files 

### ISSUE THREAD

**Issue by Max-Tait** [googleapis/nodejs-storage#2120]
#### Environment details

  - OS: Ubuntu 22.04.1 LTS
  - Node.js version: v18.12.1
  - npm version: 8,19.2
  - `@google-cloud/storage` version: 6.8.0

Hi all, I believe I have found a memory leak when downloading files.

#### Steps to reproduce
1. Create a new node project 
2. Install @google-cloud/storage
3. Replace USER_BUCKET and USER_FILE in the below code
4. Run the below code using the "--expose-gc" flag. ie "node --expose-gc  index.js"
```
"use strict";
//orginal code from https://github.com/googleapis/nodejs-storage/issues/685

const { Storage } = require("@google-cloud/storage");
const { setTimeout } = require("timers/promises");

const BASE_MEMORY = process.memoryUsage();

const USER_BUCKET = "";
const USER_FILE = "";

const storage = new Storage();
const bucket = storage.bucket(USER_BUCKET);
const file = bucket.file(USER_FILE);

async function testDownload() {
  await file.download();
  await setTimeout(30000);

  console.log("*****************Exiting*****************");
  process.exit();
}

function logMemoryUsage() {
  let curr = process.memoryUsage();

  let dif = {
    rss: Math.round((curr.rss - BASE_MEMORY.rss) / 1000000),
    heapTotal: Math.round((curr.heapTotal - BASE_MEMORY.heapTotal) / 1000000),
    heapUsed: Math.round((curr.heapUsed - BASE_MEMORY.heapUsed) / 1000000),
    external: Math.round((curr.external - BASE_MEMORY.external) / 1000000),
    arrayBuffers: Math.round(
      (curr.arrayBuffers - BASE_MEMORY.arrayBuffers) / 1000000
    ),
  };

  console.log("Memory usage:", dif);
}

setInterval(logMemoryUsage, 1000);

testDownload();
```
This code downloads a file form GCP while printing out the memory usage compared to when the program started. 

Personally I have been downloading rather large files >1gb and this is how I understand things to be working:

1. As the file downloads, the memory usage increases.
2. Once the file has finished downloading it is copied into a buffer which causes a large spike in the memory. 
3. Once garbage collection runs all fields being printed but RSS return to normal. 
4. RSS doesn't seem to ever return to the value before the file was downloaded.

![Screenshot from 2022-12-17 22-18-36](https://user-images.githubusercontent.com/19480994/208282718-39bd07e7-c8e0-4a66-963e-edb7e23904e0.png)

This is a screenshot from the ubuntu System Monitor showing the memory usage as the program is being run.

**Comment by ddelgrosso1** [comment]
Hi @Max-Tait thank you for opening an issue. I just want to clarify a few things. From your example code it looks like you are downloading the file to memory. I would expect your points 1 & 2 to be accurate in that situation. Is this issue you are seeing related to 4? I.E. you are expecting memory to return to baseline and this is where you believe there is a leak?

**Comment by Max-Tait** [comment]
Hi @ddelgrosso1, yes that correct.

**Comment by ddelgrosso1** [comment]
Thanks @Max-Tait. I will do some investigation and see if I can recreate. If you have any other logs / documentation from your testing that would also be helpful.

**Comment by ddelgrosso1** [comment]
I wanted to circle back on this. After doing some more testing and reading I'm seeing a similar behavior with regards to RSS when reading and writing a file locally using Node `fs.createReadStream` / `fs.createWriteStream`.  However, for both tests I ran, heap utilization went up and down in a pattern that would indicate space was being reclaimed.

My understanding of RSS is that it would not automatically drop unless there was demand from other processes running on the system. Are you experiencing any out of memory issues or other problems?

**Comment by ddelgrosso1** [comment]
I did some additional tests this morning to try and validate what I said in my previous comment. During this test I downloaded a 1.7GB file 5x utilizing `createReadStream` which is the underlying call `download` utilizes. 

```typescript
const s = new Storage({projectId: 'my-project'});
const b = s.bucket('my-bucket');
const f = b.file('my-large-file');

function testCreateReadStream() {
  for(let i = 0; i < 5; i++) {
    f.createReadStream()
    .pipe(fs.createWriteStream('output.txt'));
  } 
}

testCreateReadStream();
setInterval(() => {
  console.log(process.memoryUsage()['rss']);
}, 10000)
```

Once I saw that RSS had flat lined I fired up a second program that simply kept allocating memory. As more and more memory was allocated by this second program the RSS of the above began to drop. This drop became even more pronounced once the OS began to run low on available memory. The image below is a capture of my test program using `clinic`.

<img width="1451" alt="Screenshot 2023-02-01 at 11 38 44 AM" src="https://user-images.githubusercontent.com/85250797/216109941-0b95c2bf-c8a5-40fe-81d2-72f9abb00ae1.png">

I think the memory behavior you saw is not the result of anything the storage library is doing but rather how the underlying OS handles allocating / deallocating memory.
